Introduce AttributeName enumeration

cdumez opened the pull request:
#### 7279f1496728bc8835c296b8fa406a9dd110b363
<pre>
Introduce AttributeName enumeration
<a href="https://bugs.webkit.org/show_bug.cgi?id=255382">https://bugs.webkit.org/show_bug.cgi?id=255382</a>

Reviewed by Darin Adler.

Introduce AttributeName enumeration for all attribute names. We have a lot of
code that can benefit from this enumeration to do switch statements instead
of consecutive `if` checks. The new code is a bit more efficient.

This is similar to what we already did for element names with the ElementName
enumeration.

I only ported some of the code base to new AttributeName enumeration in this
patch, because the patch was getting too big. I plan to port more in follow
ups.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementsArrayReflectionAttribute):
(WebCore::Element::attributeChanged):
* Source/WebCore/dom/QualifiedName.cpp:
(WebCore::QualifiedName::QualifiedNameImpl::QualifiedNameImpl):
(WebCore::QualifiedName::QualifiedName):
(WebCore::QualifiedName::init):
(WebCore::nullQName):
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::attributeName const):
* Source/WebCore/dom/QualifiedNameCache.cpp:
(WebCore::updateImplWithNamespaceAndElementName):
(WebCore::QualifiedNameCache::getOrCreate):
* Source/WebCore/dom/QualifiedNameCache.h:
* Source/WebCore/dom/make_names.pl:
(defaultAttrPropertyHash):
(attributeCount):
(collectAllElementsAndAttrsPerNamespace):
(byAttributeNameOrder):
(printAttributeNameHeaderFile):
(printAttributeNameCppFile):
(printNamesCppFile):
(printDefinitions):
(printFactoryCppFile):
(printWrapperFactoryCppFile):
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::attributeChanged):
* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLBodyElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLElement::attributeChanged):
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::attributeChanged):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLInputElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLInputElement::attributeChanged):
* Source/WebCore/html/HTMLOListElement.cpp:
(WebCore::HTMLOListElement::attributeChanged):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::attributeChanged):
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::attributeChanged):
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::attributeChanged):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::attributeChanged):
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::qualifiedNameForTag):
(WebCore::qualifiedNameForHTMLTag):

Canonical link: <a href="https://commits.webkit.org/262929@main">https://commits.webkit.org/262929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf5902b200470a6b47588106ef8e33fa9b89816d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3054 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3119 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3439 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3023 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3193 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3151 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/2670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3084 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2740 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4263 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2718 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2702 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2769 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/4004 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3120 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2737 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2720 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/748 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2730 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2939 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/QualifiedNameCache.cpp:
```diff
@@ -65,21 +67,23 @@ Ref<QualifiedName::QualifiedNameImpl> QualifiedNameCache::getOrCreate(const Qual
-        updateImplWithNamespaceAndElementName(impl, nodeNamespace, elementName);
+        auto attributeName = AttributeName::Unknown;
+        if (elementName == ElementName::Unknown)
+            attributeName = findAttributeName(nodeNamespace, components.m_localName);
```
If we use both `findElementName` and `findAttributeName` one after another, should we find a way to collapse them into a single thing for speed?

cdumez replied:
I can try that this in a follow-up. Cameron has a similar idea. Personally, I felt it would look a little odd to have these 2 things in the same enumeration. I also don't think we construct QualifiedNames very often since they are atomized/cached and we have static ones for all well-known ones.

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/12716 (Still waiting on A/B testing results to make sure it doesn't regress Speedometer).

darinadler commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -1012,6 +1040,85 @@ sub printTagNameCppFile
+    printLicenseHeader($F);
+    print F "#pragma once\n";
+    print F "\n";
+    print F "#include \"Namespace.h\"\n";
```
Can we use a forward declaration here instead?

cdumez replied:
Probably, will try.

darinadler commented on Source/WebCore/dom/make_names.pl:
```diff
@@ -268,12 +274,18 @@ sub defaultAttrPropertyHash
     my $localName = shift;
 
     my $identifier = $localName =~ s/^x-webkit-/webkit/r =~ s/-/_/gr;
+    my $tagEnumValue = $identifier . "Attr";
```
Since these are `enum class` values, we do not need the `Attr` suffix. I think `AttributeName::src` is fine and better than `AttributeName::srcAttr`. In `HTMLNames` we used the suffix to separate tag names from attribute names and so that we could do `using namespace HTMLNames` without many conflicts.

cdumez replied:
This is actually problematic. In the enumeration, there are quite a few C++ keywords like `class` or `default`. The "Attr" suffix was a way for me to avoid this and I thought it looks OK since it was consistent with HTMLNames.

Otherwise, our perl script deals with C++ keywords by appending a `_` to the name but I thought AttributeName::class_ looked ugly.

darinadler commented on Source/WebCore/html/parser/HTMLConstructionSite.cpp:
```diff
@@ -730,15 +732,15 @@ static inline QualifiedName qualifiedNameForTag(AtomHTMLToken& token, const Atom
     if (LIKELY(elementName != ElementName::Unknown))
         return qualifiedNameForElement(elementName);
-    return { nullAtom(), token.name(), namespaceURI, nodeNamespace, elementName };
+    return { nullAtom(), token.name(), namespaceURI, nodeNamespace, elementName, AttributeName::Unknown };
```
I suspect we can just use `{ }` instead of `AttributeName::Unknown` and not have to include `AttributeName.h`. Not sure if it’s better.

cdumez committed:
Introduce AttributeName enumeration
https://bugs.webkit.org/show_bug.cgi?id=255382

Reviewed by Darin Adler.

Introduce AttributeName enumeration for all attribute names. We have a lot of
code that can benefit from this enumeration to do switch statements instead
of consecutive `if` checks. The new code is a bit more efficient.

This is similar to what we already did for element names with the ElementName
enumeration.

I only ported some of the code base to new AttributeName enumeration in this
patch, because the patch was getting too big. I plan to port more in follow
ups.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementsArrayReflectionAttribute):
(WebCore::Element::attributeChanged):
* Source/WebCore/dom/QualifiedName.cpp:
(WebCore::QualifiedName::QualifiedNameImpl::QualifiedNameImpl):
(WebCore::QualifiedName::QualifiedName):
(WebCore::QualifiedName::init):
(WebCore::nullQName):
* Source/WebCore/dom/QualifiedName.h:
(WebCore::QualifiedName::attributeName const):
* Source/WebCore/dom/QualifiedNameCache.cpp:
(WebCore::updateImplWithNamespaceAndElementName):
(WebCore::QualifiedNameCache::getOrCreate):
* Source/WebCore/dom/QualifiedNameCache.h:
* Source/WebCore/dom/make_names.pl:
(defaultAttrPropertyHash):
(attributeCount):
(collectAllElementsAndAttrsPerNamespace):
(byAttributeNameOrder):
(printAttributeNameHeaderFile):
(printAttributeNameCppFile):
(printNamesCppFile):
(printDefinitions):
(printFactoryCppFile):
(printWrapperFactoryCppFile):
* Source/WebCore/html/BaseDateAndTimeInputType.cpp:
(WebCore::BaseDateAndTimeInputType::attributeChanged):
* Source/WebCore/html/HTMLBodyElement.cpp:
(WebCore::HTMLBodyElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLBodyElement::collectPresentationalHintsForAttribute):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLElement::attributeChanged):
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::attributeChanged):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::hasPresentationalHintsForAttribute const):
(WebCore::HTMLInputElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLInputElement::attributeChanged):
* Source/WebCore/html/HTMLOListElement.cpp:
(WebCore::HTMLOListElement::attributeChanged):
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::attributeChanged):
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::attributeChanged):
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::attributeChanged):
* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::attributeChanged):
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::qualifiedNameForTag):
(WebCore::qualifiedNameForHTMLTag):

Canonical link: https://commits.webkit.org/262929@main
